File: desktop-electron/src/speech/vad.ts
/**
 * Energy-based Voice Activity Detection.
 * Direct TypeScript port of desktop/src/speech/vad.rs
 */

const FRAME_MS = 30;
const MIN_SPEECH_MS = 300;
const MAX_SPEECH_MS = 30_000;
const SILENCE_TIMEOUT_MS = 800;
const CALIBRATION_FRAMES = 66; // ~2 seconds

export type VadEvent =
  | { type: 'utterance'; samples: Float32Array }
  | { type: 'speaking-changed'; speaking: boolean }
  | { type: 'calibrated' };

export class VadProcessor {
  private frameSize: number;
  private energyThreshold = 0.01;
  private calibrationEnergies: number[] = [];
  private calibrated = false;
  private isSpeaking = false;
  private speechFrames = 0;
  private silenceFrames = 0;
  private buffer: number[] = [];
  private pending: number[] = [];

  constructor(sampleRate: number) {
    this.frameSize = Math.floor((sampleRate * FRAME_MS) / 1000);
  }

  process(samples: Float32Array | number[]): VadEvent[] {
    // Append to pending
    for (let i = 0; i < samples.length; i++) {
      this.pending.push(samples[i]);
    }

    const events: VadEvent[] = [];

    while (this.pending.length >= this.frameSize) {
      const frame = this.pending.splice(0, this.frameSize);
      this.processFrame(frame, events);
    }

    return events;
  }

  private processFrame(frame: number[], events: VadEvent[]) {
    const energy = rmsEnergy(frame);

    // Calibration phase
    if (!this.calibrated) {
      this.calibrationEnergies.push(energy);
      if (this.calibrationEnergies.length >= CALIBRATION_FRAMES) {
        const sorted = [...this.calibrationEnergies].sort((a, b) => a - b);
        const lowerHalf = sorted.slice(0, Math.floor(sorted.length / 2));
        const ambient = lowerHalf.reduce((a, b) => a + b, 0) / lowerHalf.length;
        this.energyThreshold = Math.min(Math.max(ambient * 3.0, 0.01), 0.03);
        this.calibrated = true;
        console.error(
          `[STT] VAD calibrated: ambient=${ambient.toFixed(6)}, threshold=${this.energyThreshold.toFixed(6)}`,
        );
        events.push({ type: 'calibrated' });
      }
      return;
    }

    const isSpeech = energy > this.energyThreshold;

    if (isSpeech && !this.isSpeaking) {
      console.error(`[STT] VAD: speech started (energy=${energy.toFixed(4)})`);
    }

    if (isSpeech) {
      this.silenceFrames = 0;
      this.speechFrames++;

      if (!this.isSpeaking) {
        this.isSpeaking = true;
        events.push({ type: 'speaking-changed', speaking: true });
        this.buffer.length = 0;
      }

      for (let i = 0; i < frame.length; i++) this.buffer.push(frame[i]);

      const speechMs = this.speechFrames * FRAME_MS;
      if (speechMs >= MAX_SPEECH_MS) {
        const utterance = this.finishUtterance();
        if (utterance) events.push({ type: 'utterance', samples: utterance });
        events.push({ type: 'speaking-changed', speaking: false });
      }
    } else if (this.isSpeaking) {
      this.silenceFrames++;
      for (let i = 0; i < frame.length; i++) this.buffer.push(frame[i]);

      const silenceMs = this.silenceFrames * FRAME_MS;
      if (silenceMs >= SILENCE_TIMEOUT_MS) {
        const speechMs = this.speechFrames * FRAME_MS;
        if (speechMs >= MIN_SPEECH_MS) {
          const utterance = this.finishUtterance();
          if (utterance) events.push({ type: 'utterance', samples: utterance });
        } else {
          this.resetState();
        }
        events.push({ type: 'speaking-changed', speaking: false });
      }
    }
  }

  private finishUtterance(): Float32Array | null {
    const durationMs = this.speechFrames * FRAME_MS;
    const utterance = new Float32Array(this.buffer);
    this.resetState();

    if (utterance.length === 0) return null;
    console.error(`[STT] VAD: speech ended (${durationMs}ms)`);
    return utterance;
  }

  private resetState() {
    this.isSpeaking = false;
    this.speechFrames = 0;
    this.silenceFrames = 0;
    this.buffer.length = 0;
  }
}

function rmsEnergy(frame: number[]): number {
  if (frame.length === 0) return 0;
  let sumSq = 0;
  for (let i = 0; i < frame.length; i++) sumSq += frame[i] * frame[i];
  return Math.sqrt(sumSq / frame.length);
}

File: desktop/src/speech/vad.rs
/// Simple energy-based Voice Activity Detection.
///
/// Detects speech vs silence by measuring RMS energy of audio frames.
/// When speech ends (silence exceeds threshold), the accumulated audio
/// buffer is returned as a complete utterance for transcription.

const FRAME_MS: u32 = 30; // Process in 30ms frames
const MIN_SPEECH_MS: u32 = 300; // Minimum speech duration to avoid clicks/pops
const MAX_SPEECH_MS: u32 = 30_000; // Force-segment after 30s (whisper's window)
const SILENCE_TIMEOUT_MS: u32 = 800; // 800ms silence = end of utterance
const CALIBRATION_FRAMES: usize = 66; // ~2 seconds of ambient noise calibration

/// Events produced by the VAD processor
pub enum VadEvent {
    /// A complete utterance (speech ended) — ready for transcription
    Utterance(Vec<f32>),
    /// Speaking state changed (true=started, false=ended)
    SpeakingChanged(bool),
    /// VAD calibration completed — mic is ready
    Calibrated,
}

pub struct VadProcessor {
    #[allow(dead_code)]
    sample_rate: u32,
    frame_size: usize, // samples per frame

    // Energy detection
    energy_threshold: f32,
    calibration_energies: Vec<f32>,
    calibrated: bool,

    // State
    is_speaking: bool,
    speech_frames: usize,
    silence_frames: usize,

    // Audio buffer — accumulates samples during speech
    buffer: Vec<f32>,
    // Pending samples that haven't formed a complete frame yet
    pending: Vec<f32>,
}

impl VadProcessor {
    pub fn new(sample_rate: u32) -> Self {
        let frame_size = (sample_rate * FRAME_MS / 1000) as usize;

        Self {
            sample_rate,
            frame_size,
            energy_threshold: 0.01, // Initial conservative threshold
            calibration_energies: Vec::with_capacity(CALIBRATION_FRAMES),
            calibrated: false,
            is_speaking: false,
            speech_frames: 0,
            silence_frames: 0,
            buffer: Vec::new(),
            pending: Vec::new(),
        }
    }

    /// Process incoming audio samples. Returns a list of VAD events.
    pub fn process(&mut self, samples: &[f32]) -> Vec<VadEvent> {
        self.pending.extend_from_slice(samples);

        let mut events = Vec::new();

        // Process complete frames
        while self.pending.len() >= self.frame_size {
            let frame: Vec<f32> = self.pending.drain(..self.frame_size).collect();
            self.process_frame(&frame, &mut events);
        }

        events
    }

    fn process_frame(&mut self, frame: &[f32], events: &mut Vec<VadEvent>) {
        let energy = rms_energy(frame);

        // Calibration phase: measure ambient noise
        // Uses the lower half of energy readings to ignore startup noise spikes
        if !self.calibrated {
            self.calibration_energies.push(energy);
            if self.calibration_energies.len() >= CALIBRATION_FRAMES {
                let mut sorted = self.calibration_energies.clone();
                sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
                // Use lower half median to avoid startup noise bias
                let lower_half = &sorted[..sorted.len() / 2];
                let ambient: f32 =
                    lower_half.iter().sum::<f32>() / lower_half.len() as f32;
                // 3x ambient noise, floor 0.01, ceiling 0.03
                // Floor of 0.01 prevents false triggers in quiet environments
                self.energy_threshold = (ambient * 3.0).max(0.01).min(0.03);
                self.calibrated = true;
                eprintln!(
                    "[STT] VAD calibrated: ambient={ambient:.6}, threshold={:.6}",
                    self.energy_threshold
                );
                events.push(VadEvent::Calibrated);
            }
            return;
        }

        let is_speech = energy > self.energy_threshold;

        // Log only on speech state transitions
        if is_speech && !self.is_speaking {
            eprintln!("[STT] VAD: speech started (energy={:.4})", energy);
        }

        if is_speech {
            self.silence_frames = 0;
            self.speech_frames += 1;

            if !self.is_speaking {
                self.is_speaking = true;
                events.push(VadEvent::SpeakingChanged(true));
                self.buffer.clear();
            }

            // Accumulate audio
            self.buffer.extend_from_slice(frame);

            // Force-segment if we hit max speech duration
            let speech_ms = self.speech_frames as u32 * FRAME_MS;
            if speech_ms >= MAX_SPEECH_MS {
                let utterance = self.finish_utterance();
                if let Some(u) = utterance {
                    events.push(VadEvent::Utterance(u));
                }
                events.push(VadEvent::SpeakingChanged(false));
            }
        } else if self.is_speaking {
            self.silence_frames += 1;
            // Keep buffering during short pauses
            self.buffer.extend_from_slice(frame);

            let silence_ms = self.silence_frames as u32 * FRAME_MS;
            if silence_ms >= SILENCE_TIMEOUT_MS {
                let speech_ms = self.speech_frames as u32 * FRAME_MS;
                if speech_ms >= MIN_SPEECH_MS {
                    let utterance = self.finish_utterance();
                    if let Some(u) = utterance {
                        events.push(VadEvent::Utterance(u));
                    }
                } else {
                    // Too short — probably a click or noise, discard
                    self.reset_state();
                }
                events.push(VadEvent::SpeakingChanged(false));
            }
        }
    }

    fn finish_utterance(&mut self) -> Option<Vec<f32>> {
        let utterance = std::mem::take(&mut self.buffer);
        let duration_ms = self.speech_frames as u32 * FRAME_MS;
        self.reset_state();

        if utterance.is_empty() {
            None
        } else {
            eprintln!("[STT] VAD: speech ended ({duration_ms}ms)");
            Some(utterance)
        }
    }

    fn reset_state(&mut self) {
        self.is_speaking = false;
        self.speech_frames = 0;
        self.silence_frames = 0;
    }
}

/// Root Mean Square energy of a frame
fn rms_energy(frame: &[f32]) -> f32 {
    if frame.is_empty() {
        return 0.0;
    }
    let sum_sq: f32 = frame.iter().map(|s| s * s).sum();
    (sum_sq / frame.len() as f32).sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rms_energy() {
        assert_eq!(rms_energy(&[]), 0.0);
        let energy = rms_energy(&[0.5, -0.5, 0.5, -0.5]);
        assert!((energy - 0.5).abs() < 0.001);
    }

    #[test]
    fn test_silence_no_utterance() {
        let mut vad = VadProcessor::new(16000);
        let silence = vec![0.0f32; 16000]; // 1 second of silence

        // Feed silence — should calibrate but not produce utterance
        let events = vad.process(&silence);
        // Should only have a Calibrated event, no utterances
        assert!(events.iter().all(|e| matches!(e, VadEvent::Calibrated)));
    }
}
